flink/src/test/java/org/apache/iceberg/flink/TestFlinkTableSinkV2.java

```diff
@@ -0,0 +1,333 @@
+  @Test
+  public void testChangeLogOnIdDataKey() throws Exception {
+    List<List<Row>> elementsPerCheckpoint = ImmutableList.of(
+        ImmutableList.of(
+            row("+I", 1, "aaa"),
+            row("-D", 1, "aaa"),
+            row("+I", 2, "bbb"),
+            row("+I", 1, "bbb"),
+            row("+I", 2, "aaa")
+        ),
+        ImmutableList.of(
+            row("-U", 2, "aaa"),
+            row("+U", 1, "ccc"),
+            row("+I", 1, "aaa")
+        ),
+        ImmutableList.of(
+            row("-D", 1, "bbb"),
+            row("+I", 2, "aaa")
+        )
+    );
+
+    List<List<Record>> expectedRecords = ImmutableList.of(
+        ImmutableList.of(record(1, "bbb"), record(2, "aaa"), record(2, "bbb")),
+        ImmutableList.of(record(1, "aaa"), record(1, "bbb"), record(1, "ccc"), record(2, "bbb")),
+        ImmutableList.of(record(1, "aaa"), record(1, "ccc"), record(2, "aaa"), record(2, "bbb"))
+    );
+
+    testSqlChangeLog(ImmutableList.of("data", "id"),
+        row -> Row.of(row.getField(ROW_ID_POS), row.getField(ROW_DATA_POS)),
+        elementsPerCheckpoint, expectedRecords);
+  }
+
+  @Test
+  public void testPureInsertOnIdKey() throws Exception {
+    List<List<Row>> elementsPerCheckpoint = ImmutableList.of(
+        ImmutableList.of(
+            row("+I", 1, "aaa"),
+            row("+I", 2, "bbb")
+        ),
+        ImmutableList.of(
+            row("+I", 3, "ccc"),
+            row("+I", 4, "ddd")
+        ),
+        ImmutableList.of(
+            row("+I", 5, "eee"),
+            row("+I", 6, "fff")
+        )
+    );
+
+    List<List<Record>> expectedRecords = ImmutableList.of(
+        ImmutableList.of(
+            record(1, "aaa"),
+            record(2, "bbb")
+        ),
+        ImmutableList.of(
+            record(1, "aaa"),
+            record(2, "bbb"),
+            record(3, "ccc"),
+            record(4, "ddd")
+        ),
+        ImmutableList.of(
+            record(1, "aaa"),
+            record(2, "bbb"),
+            record(3, "ccc"),
+            record(4, "ddd"),
+            record(5, "eee"),
+            record(6, "fff")
+        )
+    );
+
+    testSqlChangeLog(ImmutableList.of("data"), row -> Row.of(row.getField(ROW_DATA_POS)), elementsPerCheckpoint,
+        expectedRecords);
+  }
+
+  private static Row row(String rowKind, int id, String data) {
+    return SimpleDataUtil.createRow(rowKind, id, data);
+  }
+
+  private Record record(int id, String data) {
+    return SimpleDataUtil.createRecord(id, data);
+  }
+
+  private void testSqlChangeLog(List<String> equalityColumns,
+                                KeySelector<Row, Row> keySelector,
+                                List<List<Row>> inputRowsPerCheckpoint,
+                                List<List<Record>> expectedRecordsPerCheckpoint) throws Exception {
+    String partitionByCause = partitioned ? "PARTITIONED BY (data)" : "";
+    sql("CREATE TABLE %s (id INT, data VARCHAR, PRIMARY KEY(%s) NOT ENFORCED) %s WITH %s",
```

Comment:
Instead of creating v2 tables using the table property, I think it would be better to load the table just after creation and update it using the API. That way you can test with a v2 table, but we don't need to change any of the core classes or expose the format before it is finalized.

Reply:
OK,  I saw there's an `upgradeToFormatVersion` API  to do the upgrade  in [TestSparkReaderDeletes.java](https://github.com/apache/iceberg/blob/master/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java#L101),  I can use the similar way.